design/authn-oidc.md

```diff
@@ -0,0 +1,257 @@
+  - [Security](#security)
+  - [Audit](#audit)
+  - [Development Tasks](#development-tasks)
+  - [Solution Review](#solution-review)
+
+## Useful Links
+
+| Link | Private |
+|------|:-------:|
+| [Conjur Docs: OIDC Authentication](https://docs.conjur.org/Latest/en/Content/OIDC/OIDC.htm) | No |
+| [Solution Design: Authn-LDAP Support in CLI](authn-ldap.md) | No |
+| [MFA for Conjur UI - Architecture](https://ca-il-confluence.il.cyber-ark.com/display/rndp/MFA+for+Conjur+UI+-+Architecture) | Yes |
+| [MFA Customer Documentation Changes](https://ca-il-confluence.il.cyber-ark.com/display/rndp/MFA+Customer+Documentation+Changes+-+Dev+Content) | Yes |
+
+## Overview
+
+Conjur supports various forms of authentication, including OIDC (OpenID Connect). OIDC is unique in that requires the user to
+navigate to a third party website (SSO provider) and authenticate there, often using MFA. The user then receives a token from
+the SSO provider and passes this to Conjur, which verifies that the token is signed correctly by the trust SSO provider.
+This document describes the design and implementation of the OIDC authentication support in the new Python-based Conjur CLI.
+
+### Motivation
+
+OIDC support was recently added to the Conjur UI (Enterprise only), initially to support SSO with Okta. We want to extend this
+support to the CLI so developers using the CLI can authenticate to Conjur in the CLI with OIDC as they will in the Conjur UI.
+
+## Solution
+
+### User Experience
+
+The `conjur init` command should support OIDC authentication as an authentication type. It should be possible to specify
+the authentication type `oidc` using the existing `-t` or `--authn-type` option.
+When using `oidc`, a `--service-id` option should be mandatory.
+
+*Open question*: Currently, if the `--service-id` option is specified, then `--authn-type` defaults to `ldap`. Should we make this required
+without a default value instead?
+
+| Option | Acceptable Values | Default | Description |
+| ------ | ----------------- | ------- | ----------- |
+| `--authn-type` / `-t` | `authn`, `ldap`, `oidc` | ??? | The authentication method to use when connecting to Conjur |
+| `--service-id` | `<service_id>` | N/A | The service_id of the OIDC service to use |
+
+The `conjur login` command should open a browser (and print a link) to the OIDC URL configured for the SSO provider.
+The user must then authenticate to the SSO provider and receive a JWT which will be stored in the user's keychain or in the .netrc file.
+(See [below](#in-the-api-repository) for how the CLI will get the JWT token from the browser.)
+Then when executing any CLI commands, the CLI will retrieve the stored JWT and pass it to Conjur's OIDC authentication endpoint and receive
+a short-lived accesss token, as with all other authentication strategies.
+
+*Open question*: Can we revoke the JWT when `conjur logout` is called?
+
+#### Feature Details
+
+##### Diagram
+
+![Diagram](assets/authn-oidc.png)
```

Comment:
I'm still a bit confused by our usage of `redirect_uri` in this context.  Usually in OAuth/OIDC the redirect_uri refers to the endpoint which the OIDC provider redirects the authorization code and state/nonce data to **after a successful authentication attempt**. Here it seems like we are calling the Okta authorization endpoint that gets opened on the client browser a `redirect_uri`.  Later in the diagram (and this document) we are referring to the redirect_uri that Okta uses as a callback_url.

In my mind, the term `redirect_uri should be reserved for the required parameter sent to Okta in the authentication request [per the ODIC spec](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest). 

Example (assuming ad-hoc HTTP server):
redirect_uri = https://localhost:{port}/handleOidcRedirect

Authorization endpoint = https://cyberark.okta.com?client_id=<client_id>&redirect_uri=<redirect_uri>&response_type=code&scope=openid%2fprofile%2femail&state=<state>&nonce=<nonce>

Maybe @jvanderhoof can weigh in on this since I think this diagram is based on the ConjurUI design.

Reply:
Ah sorry - I updated the language in the markdown but forgot to in the diagram. I'll fix this now.

Reply:
It's true that it's called "redirect_uri" in the Conjur UI design doc but that's probably because from the perspective of the UI which is a web app, it's the URL that the user will be *redirected* to. I'm going to try to make it a bit clearer in the diagram here.

Reply:
@gl-johnson Diagram updated. Please let me know if that's clearer.

Reply:
Yep - thanks for addressing